api/clients/v2/accountant.go

```diff
@@ -67,55 +70,77 @@ func NewAccountant(accountID gethcommon.Address, reservation *core.ReservedPayme
 // indicating on-demand payment.
 // These generated values are used to create the payment header and signature, as specified in
 // api/proto/common/v2/common_v2.proto
-func (a *Accountant) blobPaymentInfo(
-	numSymbols uint64,
+// func (a *Accountant) blobPaymentInfo(
+// 	numSymbols uint64,
+// 	quorumNumbers []uint8,
+// 	timestamp int64) (*big.Int, error) {
+
+// 		symbolUsage := meterer.SymbolsCharged(numSymbols, a.minNumSymbols)
+
+// 		// Always try to use reservation first
+// 		err := a.reservationUsage(symbolUsage, quorumNumbers, timestamp)
+// 		if err == nil {
+// 			return big.NewInt(0), nil
+// 		}
+
+// 		// Fall back to on-demand payment if reservation fails
+// 		return a.onDemandUsage(symbolUsage, quorumNumbers)
+// }
+
+// reservationUsage attempts to use the reservation for the given request.
+func (a *Accountant) reservationUsage(
+	symbolUsage uint64,
 	quorumNumbers []uint8,
-	timestamp int64) (*big.Int, error) {
+	timestamp int64) error {
+	if err := meterer.ValidateQuorum(quorumNumbers, a.reservation.QuorumNumbers); err != nil {
+		return err
+	}
+
 	reservationWindow := a.reservationWindow
 	currentReservationPeriod := meterer.GetReservationPeriodByNanosecond(timestamp, reservationWindow)
-	symbolUsage := a.symbolsCharged(numSymbols)
 
-	a.usageLock.Lock()
-	defer a.usageLock.Unlock()
+	a.periodRecordsLock.Lock()
+	defer a.periodRecordsLock.Unlock()
 	relativePeriodRecord := a.getOrRefreshRelativePeriodRecord(currentReservationPeriod, reservationWindow)
 	relativePeriodRecord.Usage += symbolUsage
 
-	// first attempt to use the active reservation
-	binLimit := a.reservation.SymbolsPerSecond * uint64(a.reservationWindow)
+	// Check if we can use the reservation within the bin limit
+	binLimit := meterer.GetReservationBinLimit(a.reservation, a.reservationWindow)
 	if relativePeriodRecord.Usage <= binLimit {
-		if err := QuorumCheck(quorumNumbers, a.reservation.QuorumNumbers); err != nil {
-			return big.NewInt(0), err
-		}
-		return big.NewInt(0), nil
+		return nil
 	}
 
 	overflowPeriodRecord := a.getOrRefreshRelativePeriodRecord(currentReservationPeriod+2*reservationWindow, reservationWindow)
```

Comment:
Sorry to add more scope to the PR but I feel the math here `currentReservationPeriod+2*reservationWindow` should be documented somewhere, a comment explaining the logic should be sufficient. Also this `2 * reservationWindow` should be some variable name. Maybe `currentReservationPeriod +  overflowPeriodOffset`?